src/shop-checkout.js

```diff
@@ -514,25 +507,17 @@ class ShopCheckout extends Element {
    * Sets the valid state and updates the location.
    */
   _pushState(state) {
-    this._validState = state;
-    this.set('route.path', state);
-  }
-
-  /**
-   * Checks that the `:state` subroute is correct. That is, the state has been pushed
-   * after receiving response from the server. e.g. Users can only go to `/checkout/success`
-   * if the server responsed with a success message.
-   */
-  _updateState(active, routeData) {
-    if (active && routeData) {
-      let state = routeData.state;
-      if (this._validState === state) {
-        this.state = state;
-        this._validState = '';
-        return;
-      }
-    }
-    this.state = 'init';
+    // This changes window.location only - it does not affect the checkout state.
+    window.history.pushState({}, '', `${window.location.origin}/checkout/${state}`);
+    window.dispatchEvent(new CustomEvent('location-changed'));
```

Comment:
This stuff should probably follow the same action dispatch / state flow as the rest of the app. Consider encapsulating history transitions into a middleware.